js/plugins/vertexai/src/vector-search/types.ts

```diff
@@ -0,0 +1,192 @@
+      ]),
+      z.null(),
+    ])
+    .optional(),
+});
+
+export type NumericRestriction = z.infer<typeof NumericRestrictionSchema>;
+
+// Define the Zod schema for ICrowdingTag
+export const CrowdingTagSchema = z.object({
+  crowdingAttribute: z.string().optional(),
+});
+
+export type CrowdingTag = z.infer<typeof CrowdingTagSchema>;
+
+// Define the Zod schema for IIndexDatapoint
+const IndexDatapointSchema = z.object({
+  datapointId: z.string().optional(),
+  featureVector: z.array(z.number()).optional(),
+  sparseEmbedding: SparseEmbeddingSchema.optional(),
+  restricts: z.array(RestrictionSchema).optional(),
+  numericRestricts: z.array(NumericRestrictionSchema).optional(),
+  crowdingTag: CrowdingTagSchema.optional(),
+});
+
+// Define the Zod schema for INeighbor
+export const NeighborSchema = z.object({
+  datapoint: IndexDatapointSchema.optional(),
+  distance: z.number().optional(),
+  sparseDistance: z.number().optional(),
+});
+
+export type Neighbor = z.infer<typeof NeighborSchema>;
+
+// Define the Zod schema for INearestNeighbors
+const NearestNeighborsSchema = z.object({
+  id: z.string().optional(),
+  neighbors: z.array(NeighborSchema).optional(),
+});
+
+// Define the Zod schema for IFindNeighborsResponse
+export const FindNeighborsResponseSchema = z.object({
+  nearestNeighbors: z.array(NearestNeighborsSchema).optional(),
+});
+
+export type FindNeighborsResponse = z.infer<typeof FindNeighborsResponseSchema>;
+
+// TypeScript types for Zod schemas
+type IndexDatapoint = z.infer<typeof IndexDatapointSchema>;
+
+// Function to assert type equality
+function assertTypeEquality<T>(value: T): void {}
+
+// Asserting type equality
+assertTypeEquality<IIndexDatapoint>({} as IndexDatapoint);
+assertTypeEquality<IFindNeighborsResponse>({} as FindNeighborsResponse);
+
+export const VertexAIVectorRetrieverOptionsSchema =
+  CommonRetrieverOptionsSchema.extend({}).optional();
+
+export type VertexAIVectorRetrieverOptions = z.infer<
+  typeof VertexAIVectorRetrieverOptionsSchema
+>;
+
+export const VertexAIVectorIndexerOptionsSchema = z.any();
+
+export type VertexAIVectorIndexerOptions = z.infer<
+  typeof VertexAIVectorIndexerOptionsSchema
+>;
+
+/**
+ * A document retriever function that takes an array of Neighbors from Vertex AI Vector Search query result, and resolves to a list of documents.
+ * Also takes an options object that can be used to configure the retriever.
+ */
+export type DocumentRetriever<Options extends { k?: number } = { k?: number }> =
+  (docIds: Neighbor[], options?: Options) => Promise<Document[]>;
+
+/**
+ * Indexer function that takes an array of documents, stores them in a database of the user's choice, and resolves to a list of document ids.
+ * Also takes an options object that can be used to configure the indexer.
+ */
+export type DocumentIndexer<Options extends {} = {}> = (
+  docs: Document[],
+  options?: Options
+) => Promise<string[]>;
+
+export interface VectorSearchIndexOption<
```

Comment:
Optional: Have you considered naming this `VectorSearchOptions`? The inclusion of `Index` might suggest that this is only relevant for indexing and not retrieving. Although Index here means a DB index I suppose (rather than the act of indexing). 

WDYT?

cc: @pavelgj